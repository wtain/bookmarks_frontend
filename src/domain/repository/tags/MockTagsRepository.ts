import ITagsRepository from './ITagsRepository';

class MockTagsRepository implements ITagsRepository {
  async getTags(): Promise<string[]> {
    return ["tag1", "tag2", "tag3"];
  }
  
}

export default MockTagsRepository;