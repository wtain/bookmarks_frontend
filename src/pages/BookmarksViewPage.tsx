
import useEventListener from "@use-it/event-listener";
import { useEffect, useState } from "react";
import BookmarkForm from "../components/UI/BookmarkForm/BookmarkForm";
import BookmarkList from "../components/UI/BookmarkList/BookmarkList";
import Loading from "../components/UI/Loading/Loading";
import Modal from "../components/UI/Modal/Modal";
import BookmarkDto from "../domain/dto/BookmarkDto";
import TagDto from "../domain/dto/TagDto";
import IBookmarksRepository from "../domain/repository/bookmarks/IBookmarksRepository";
import ITagsRepository from "../domain/repository/tags/ITagsRepository";
import cl from './BookmarksPage.module.css'
import BookmarksFilterDto, { EmptyFilter } from '../domain/dto/BookmarksFilterDto';
import FilterResultDto from "../domain/dto/FilterResultDto";
import BookmarksFilter from "../components/UI/BookmarksFilter/BookmarksFilter";

interface Props {
    bookmarksRepository: IBookmarksRepository;
    tagsRepository: ITagsRepository;
}

// Copy of BookmarksPage
const BookmarksViewPage = (props: Props) => {

    const [modal, setModal] = useState<boolean>(false);

    const [newBookmarkId, setNewBookmarkId] = useState<string | undefined>("");

    const [bookmarks, setBookmarks] = useState<BookmarkDto[]>([]);
  
    const [filter, setFilter] = useState<BookmarksFilterDto>(EmptyFilter);

    const [loading, setLoading] = useState<boolean>(false);

    const getData = async (): Promise<FilterResultDto> => {
      return await props.bookmarksRepository.filterBookmarks(filter);
    }

    const loadBookmarks = async (success: (bookmarks: BookmarkDto[]) => void, error: (e: unknown) => void) => {
        try {
            const bookmarks = await getData();
            success(bookmarks.bookmarks);
        } catch (e) {
            error(e);
        }
    }

    useEffect(() => {
        setLoading(true);
        loadBookmarks((bookmarks: BookmarkDto[]) => {
            setBookmarks(bookmarks)
            setLoading(false);
        }, 
          (_e) => {
              // todo: indicate error state, show error popup
                setLoading(false);
            })
    }, [filter])

    const doUpdatePoll = async () => {
        const bookmarks = await getData();
        setBookmarks(bookmarks.bookmarks);
    }

    const addBookmark = async (newBookmark: BookmarkDto) => {
        setModal(false);
        setNewBookmarkId(newBookmark.id);
        await props.bookmarksRepository.addBookmark(newBookmark);
        doUpdatePoll();
    }

    const removeBookmark = async (bm: BookmarkDto) => {
        setNewBookmarkId("");
        await props.bookmarksRepository.removeBookmark(bm);
        doUpdatePoll();
    }

    useEffect(() => {
        const interval = setInterval(() => doUpdatePoll(), 20 * 1000);
        return () => clearInterval(interval);
    }, [filter]);

    useEventListener('keydown', (e: KeyboardEvent) => {
        if (e.key === "a" && e.ctrlKey) {
            setModal(true);
            e.stopPropagation();
        } else if (e.key === "Escape" && modal) {
            setModal(false);
        }
    })
  
  // todo: Filter: add/remove tags, set dates

    return (
        <div>
            <Modal visible={modal} setVisible={setModal}>
                <BookmarkForm 
                        create={addBookmark} 
                        visibility={modal}
                        validation={(bookmark: BookmarkDto) => bookmark.contents.trim().length > 0 && bookmark.summary.trim().length > 0} />
            </Modal>
            <button onClick={() => setModal(true)} className={cl.btn_add}>
            âž•
            </button>

            <BookmarksFilter tagsRepository={props.tagsRepository} onFilterChanged={
                (filter: BookmarksFilterDto) => {
                    setFilter(filter);
                }
            } />
        

            {
                loading ? 
                <Loading /> 
                : 
                <BookmarkList bookmarks={bookmarks} 
                        newBookmarkId={newBookmarkId}
                        onRemoveBookmark={(bm: BookmarkDto) => removeBookmark(bm)} 
                        onBookmarkContentsChanged={async (bm: BookmarkDto, new_contents: string) => {
                            await props.bookmarksRepository.editBookmark({...bm, contents: new_contents, updated: new Date()});
                            doUpdatePoll();
                        }}
                        onBookmarkSummaryChanged={async (bm: BookmarkDto, new_summary: string) => {
                            await props.bookmarksRepository.editBookmark({...bm, summary: new_summary, updated: new Date()});
                            doUpdatePoll();
                        }}
                        onBookmarkTagAdded={async (bm: BookmarkDto, new_tag: TagDto) => {
                            await props.bookmarksRepository.editBookmark({...bm, tags: [...bm.tags, new_tag], updated: new Date()});
                            doUpdatePoll();
                        }}
                        onBookmarkTagRemoved={async (bm: BookmarkDto, index: number) => {
                            await props.bookmarksRepository.editBookmark({...bm, tags: bm.tags.filter((v, i) => i !== index), updated: new Date()});
                            doUpdatePoll();
                        }}
                        onBookmarkIsDoneChanged={async (bm: BookmarkDto, new_value: boolean) => {
                            await props.bookmarksRepository.editBookmark({...bm, isDone: new_value, updated: new Date()});
                            doUpdatePoll();
                        }}
                    />
            }
            
        </div>
    )
}

export default BookmarksViewPage;