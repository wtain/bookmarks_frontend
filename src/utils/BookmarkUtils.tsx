
import React from "react";
import BookmarkDto from "../domain/dto/BookmarkDto";
import {v4 as uuidv4} from 'uuid';
import TagDto from "../domain/dto/TagDto";

const createBookmark = (summary: string, contents: string, tags: TagDto[]): BookmarkDto => {
    return {
                summary, 
                contents, 
                id: uuidv4(), 
                created: new Date().toLocaleString(),
                tags,
                isDone: false
            }
}

const createNewTag = (): TagDto => {
    return { name: "", id: uuidv4() }
}

const BookmarkUtils = { createBookmark, createNewTag };

export default BookmarkUtils;