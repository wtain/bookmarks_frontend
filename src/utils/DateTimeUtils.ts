import moment from "moment";

export function parseDateTimeFromMongo(datetime: Date | undefined) {
  if (datetime === undefined) {
    return undefined;
  }
  return moment(datetime.toISOString()).fromNow();
}

export function addOneMonth(date: Date): Date {
  let result = new Date(date);
  result.setMonth(result.getMonth() + 1)
  return result;
}

export function firstDayOfMonth(date: Date): Date {
  let result = new Date(date);
  result.setDate(1)
  return result;
}

export function getDatePortion(date: Date): string {
  const dateCopy = new Date(date);
  dateCopy.setUTCHours(0);
  dateCopy.setMinutes(0);
  const momentForDate = moment(dateCopy);
  return momentForDate.format('YYYY-MM-DD');
}

export function datesEqual(date1: Date, date2: Date): boolean {
  return getDatePortion(date1) === getDatePortion(date2);
}