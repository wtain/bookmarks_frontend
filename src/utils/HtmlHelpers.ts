import sanitizeHtml from "sanitize-html";

export function sanitizeText(text: string){
  return sanitizeHtml(text, {
    allowedTags: ['br', 'b', 'a', 'img']
  });
}