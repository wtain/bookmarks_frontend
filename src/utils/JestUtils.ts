
export function jestHacks() {
    // https://stackoverflow.com/questions/39830580/jest-test-fails-typeerror-window-matchmedia-is-not-a-function
    // https://github.com/facebook/create-react-app/issues/10126
    window.matchMedia = (query) => ({
        matches: false,
        media: query,
        onchange: null,
        addListener: jest.fn(), // Deprecated
        removeListener: jest.fn(), // Deprecated
        addEventListener: jest.fn(),
        removeEventListener: jest.fn(),
        dispatchEvent: jest.fn(),
      })
}