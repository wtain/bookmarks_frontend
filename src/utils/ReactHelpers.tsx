
import React, { MutableRefObject } from "react";

const scrollToRef = (ref: MutableRefObject<null | HTMLDivElement>) => {
    if (ref.current) {
        ref.current!.scrollIntoView({
            behavior: "smooth",
          });
        // On mobile Chrome smooth is not supported, so let's 
        // make sure we're scrolled anyway
        // ref.current!.scrollIntoView(true);
    }
};

const delay = (milliseconds: number) => {
    return new Promise(resolve => setTimeout(resolve, milliseconds));
}

const ReactHelpers = {
    scrollToRef, delay
}

export default ReactHelpers;