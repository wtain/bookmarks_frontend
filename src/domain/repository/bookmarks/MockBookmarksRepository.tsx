
import React from "react";
import BookmarkDto from "../../dto/BookmarkDto";
import IBookmarksRepository from "./IBookmarksRepository";
import ReactHelpers from "../../../utils/ReactHelpers"
import BookmarkUtils from "../../../utils/BookmarkUtils";

class MockBookmarksRepository implements IBookmarksRepository {

    private bookmarks: BookmarkDto[];

    constructor() {
        this.bookmarks = Array.from(Array(20).keys()).map(i => 
            BookmarkUtils.createBookmark("Summary " + i, 
                "Contents " + i + "\nLorem ipsum\nLorem ipsum", 
                Array.from(Array(i).keys()).map(j => { 
                    return { ...BookmarkUtils.createNewTag(), name: "tag " + j }
                })
            ));
    }

    async getBookmark(id: string): Promise<BookmarkDto> {
        // todo: implement
        await ReactHelpers.delay(300);
        return this.bookmarks[0];
    }

    public async getBookmarks(): Promise<BookmarkDto[]> {
        await ReactHelpers.delay(300);
        return this.bookmarks;
    }

    async getBookmarksByTag(tag: string): Promise<BookmarkDto[]> {
        await ReactHelpers.delay(300);
        return this.bookmarks; // todo: implement
    }

    public async addBookmark(bookmark: BookmarkDto) {
        this.bookmarks = [...this.bookmarks, bookmark]
    }

    public async removeBookmark(bookmark: BookmarkDto) {
        this.bookmarks = this.bookmarks.filter(b => b.id != bookmark.id);
    }

    public async editBookmark(bookmark: BookmarkDto) {
        this.bookmarks = this.bookmarks.map(b => b.id === bookmark.id ? bookmark : b);
    }
}

export default MockBookmarksRepository;