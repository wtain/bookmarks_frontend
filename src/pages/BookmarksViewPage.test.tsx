import renderer from 'react-test-renderer';
import MockBookmarksRepository from '../domain/repository/bookmarks/MockBookmarksRepository';
import MockTagsRepository from '../domain/repository/tags/MockTagsRepository';
import { BrowserRouter } from 'react-router-dom';
import BookmarksViewPage from './BookmarksViewPage';
import { jestHacks } from '../utils/JestUtils';
import { AppProvider } from "@shopify/polaris";
import translations from "@shopify/polaris/locales/en.json";

describe('BookmarksViewPage tests', () => {
  beforeAll(() => {
    jestHacks();
  });

  it('BookmarksViewPage renders correctly', () => {
    const bookmarksRepository = new MockBookmarksRepository();
    const tagsRepository = new MockTagsRepository();

    const tree = renderer
      .create(
        <BrowserRouter>
          <AppProvider i18n={translations}>
            <BookmarksViewPage bookmarksRepository={bookmarksRepository} tagsRepository={tagsRepository} />
          </AppProvider>
        </BrowserRouter>
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
});