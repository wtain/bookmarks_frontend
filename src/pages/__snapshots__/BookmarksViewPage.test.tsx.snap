// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BookmarksViewPage renders correctly 1`] = `
<div>
  <div
    className="Modal"
    onClick={[Function]}
  >
    <div
      className="ModalContent"
      onClick={[Function]}
    >
      <form
        className="bookmark_form form_invalid"
      >
        <input
          className="input"
          onChange={[Function]}
          placeholder="Summary"
          value=""
        />
        <div
          className="textarea_wrapper"
        >
          <textarea
            className="textarea"
            onChange={[Function]}
            onKeyDown={[Function]}
            placeholder="Contents"
            rows={3}
            value=""
          />
        </div>
        <div
          className="container tags_container"
          onClick={[Function]}
        >
          <span
            className="sp"
          >
            <input
              className="input"
              onBlur={[Function]}
              onChange={[Function]}
              onKeyDown={[Function]}
              placeholder="Tags"
              value=""
            />
          </span>
        </div>
        <button
          className="button"
          onClick={[Function]}
        >
          ✔️Add
        </button>
      </form>
    </div>
  </div>
  <button
    className="btn_add"
    onClick={[Function]}
  >
    ➕
  </button>
  <div
    className="filter"
  >
    <span
      className="tag"
      onClick={[Function]}
      onMouseEnter={[Function]}
      onMouseLeave={[Function]}
    >
      ❌
    </span>
    <input
      checked={true}
      onChange={[Function]}
      type="checkbox"
    />
    All tags?
    <select
      onChange={[Function]}
      value={0}
    >
      <option
        value={0}
      >
        All
      </option>
      <option
        value={1}
      >
        Todo
      </option>
      <option
        value={2}
      >
        Done
      </option>
    </select>
  </div>
  <div>
    (No bookmarks)
  </div>
</div>
`;
